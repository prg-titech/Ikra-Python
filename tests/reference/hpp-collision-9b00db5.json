{
    "basename": "hpp-collision-9b00db5",
    "cmd": "new-src/interface.py --emit-hpp {infile}",
    "infile": "examples/collision/collision.py",
    "infile_hash": "32b6014a5cbf9597cb4b52c12e91d20c646ad1a3138607b7120da6e7",
    "stdout": "hpp-collision-9b00db5.stdout",
    "stdout_hash": "c6efe07a87fa5688613e684b0e4156b3e97f470d433cba18148a6fc3",
    "stderr": null,
    "stderr_hash": null,
    "returncode": 0
}