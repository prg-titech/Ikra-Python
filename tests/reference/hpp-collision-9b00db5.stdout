#ifndef SANAJEH_DEVICE_CODE_H
#define SANAJEH_DEVICE_CODE_H
#define KNUMOBJECTS 64*64*64*64

#include <curand_kernel.h>
#include "dynasoar.h"

class Body;

using AllocatorT = SoaAllocator<KNUMOBJECTS, Body>;

static const int kSeed = 42;
static const float kMaxMass = 500;
static const float kGravityConstant = 4e-06;
static const float kDampeningFactor = 0.05;
static const float kMergeThreshold = 0.1;
static const float kTimeInterval = 0.05;

class Body : public AllocatorT::Base 
{
public:
	declare_field_types(Body, curandState, Body*, float, float, float, float, float, float, float, bool, bool)
	Field<Body, 0> random_state_;
	Field<Body, 1> merge_target_ref;
	Field<Body, 2> pos_x;
	Field<Body, 3> pos_y;
	Field<Body, 4> vel_x;
	Field<Body, 5> vel_y;
	Field<Body, 6> force_x;
	Field<Body, 7> force_y;
	Field<Body, 8> mass;
	Field<Body, 9> has_incoming_merge;
	Field<Body, 10> successful_merge;
	
	__device__ Body(int idx);
	__device__ void Body__init(int idx);
	__device__ void compute_force();
	__device__ void apply_force(Body* other);
	__device__ void body_update();
	__device__ void check_merge_into_this(Body* other);
	__device__ void initialize_merge();
	__device__ void prepare_merge();
	__device__ void update_merge();
	__device__ void delete_merged();
	__device__ Body() {};
	void _do(void (*pf)(int, int, float, float, float, float, float, float, float, bool, bool));
};

extern "C" int Body_do_all(void (*pf)(int, int, float, float, float, float, float, float, float, bool, bool));
extern "C" int Body_Body_compute_force();
extern "C" int Body_Body_body_update();
extern "C" int Body_Body_initialize_merge();
extern "C" int Body_Body_prepare_merge();
extern "C" int Body_Body_update_merge();
extern "C" int Body_Body_delete_merged();
extern "C" int parallel_new_Body(int object_num);
extern "C" int AllocatorInitialize();
extern "C" int AllocatorUninitialize();

#endif
